### PR TITLE
feat: Trials Comparison Frontend

### PR DESCRIPTION
**Pull request by trentwatt**
## Description

[WEB-487](https://determinedai.atlassian.net/browse/WEB-487) follow on to #4543.

brings the work on trials comparison frontend into the master branch.

the tab itself remains hidden, you actually have to go to e.g. http://localhost:3000/det/projects/2/experiments?f_trials_comparison=on for it to show up.

i ask that the criteria for merging be that we feel okay about the impact and changes to existing code, and that we leave considerations about the actual feature itself to follow up work (that would happen before making it visible).

#### Test: Existing Behavior is Preseved
**all experiment list functionality**
- [ ] sorting
    - [ ] asc/desc by default column (start time)
    - [ ] asc/desc both other columns
- [ ] filtering
    - [ ] by state
    - [ ] by experiment name
    - [ ] by tags
- [ ] navigation
    - [ ] to and from project details
    - [ ] to and from notes tab of project details
    - [ ] using browser back and forward for the above navigations 
    - [ ] refreshing the page

#### Test: New Functionality 
to verify that it works as intended

**basic**   
- [ ] start a project and run a bunch of single and multi-trial experiments in it
- [ ] go to http://{cluster_address}/det/projects/{project_id}/experiments?f_trials_comparison=on and then navigate to trials tab
    - [ ] you will see some button/dropdowns in the top right of the screen
        - [ ] this is the interface for interacting with trials "collections" (name is WIP)
    - [ ] you will also see a gallery of graphs of trials in the top half of the screen and a table of trials at the bottom
        - [ ] if you make a selection, only the selected trials will appear
        - [ ] clicking and dragging to zoom on an individual chart will apply to all of the charts
        - [ ] you can apply change filters and sort the table, and the relevant trials will be displayed above
        - [ ] when you navigate to and from the page or tab, your the view should be preserved
        - [ ] you can view the currently active filter(s) and and sorter by clicking on the button in the top right with the three horizontal lines  

    - [ ] **collections**:  you can also save a current view as a "collection." to do so, click the "new collection" button, enter a name, and press confirm.
        - [ ] you should now see the name of the collection you created selected in the dropdown
        - [ ] there should be a person icon next to the name indicating that it was created by you
        - [ ] at this point, you can continue modifying the filters and sorter. if the filters/sorter are modified, you should the collection selector glow with a faint orange highlight. 
        - [ ] you can also check again to see what the modified filters/sorter are
        - [ ] if you navigate away from the page/tab, your filters should still be modified as you last left them, and there should still be the orange glow
        - [ ] if you wish, you can reset the filters/sorter to those of the collection by clicking the reset button
        - [ ] alternatively, you can save the changes to the filter/sorter to the collection by clicking the button with the checkmark  

    - [ ] **collection interactions**: you can also create a separate collection for a different set of filters/sorter
        - [ ] when you do, you can flip back and for between the corresponding filters/sorter
            - [ ] the relevant filters/sorter should be applied when you do so
            - [ ] if you have made any modifications to the current filter/sorter state, it will be cleared  
    - [ ] **additional table functionality** : to test table actions, you can begin by making a selection, and then the actions menu will show up
        - [ ] add tags will add tags to the selected trials
        - [ ] tag & collect will create a collection with the name you specify, that contains all trials that are tagged with that same name, and also give that tag to the selected trials
        - [ ] there is also the option to "select all matching" this will apply the action (in this case, only tagging) to all trials that match the current filters
    - [ ] test sorting/filtering by any/all available fields
       - [ ] the "Exp Rank" control allows you to sort/filter by trials according to their rank within their respective experiments
  

thanks for testing!

## Commentary (optional)

I will be following up with comments throughout the code itself to call out various things.

Also down for pairing sessions to review together.

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.
- [x] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/pages/TrialDetails/TrialDetailsWorkloads.tsx [hunk 1]
```diff
@@ -42,16 +41,16 @@ export interface Props {
 const TrialDetailsWorkloads: React.FC<Props> = ({
   defaultMetrics,
   experiment,
-  metrics,
+  selectedMetrics,
```
Comment (trentwatt) [inline webui/react/src/pages/TrialDetails/TrialDetailsWorkloads.tsx:44]: there should be `allMetrics` here or similar (and it should be used below)

File: webui/react/src/pages/TrialDetails/TrialDetailsOverview.tsx [hunk 2]
```diff
@@ -79,8 +80,7 @@ const TrialDetailsOverview: React.FC<Props> = ({ experiment, trial }: Props) =>
       <TrialDetailsWorkloads
         defaultMetrics={defaultMetrics}
         experiment={experiment}
-        metricNames={metricNames}
-        metrics={metrics}
+        selectedMetrics={metrics}
         settings={settings}
```
Comment (trentwatt) [inline webui/react/src/pages/TrialDetails/TrialDetailsOverview.tsx:84]: there should be `allMetrics` here or similar (and it should be used in the component)

**Comment by ClaireNeveu** [comment]
Can you add a feature switch and show the tab when the feature switch is enabled?

**Comment by trentwatt** [comment]
> Can you add a feature switch and show the tab when the feature switch is enabled?

ok done

File: webui/react/src/services/apiConfig.ts [hunk 3]
```diff
@@ -435,6 +438,92 @@ export const getResourceAllocationAggregated: DetApi<
...
+    return detApi.TrialsComparison.getTrialsCollections(projectId);
+  },
+};
+
+export const patchTrialsCollection: DetApi<
+  Api.V1PatchTrialsCollectionRequest,
+  Api.V1PatchTrialsCollectionResponse,
+  TrialsCollection | undefined
+> = {
+  name: 'getTrialsCollection',
```
Comment (julian-determined-ai) [inline webui/react/src/services/apiConfig.ts:507]:
```suggestion
  name: 'patchTrialsCollection',
```

File: webui/react/src/services/apiConfig.ts [hunk 4]
```diff
@@ -435,6 +438,92 @@ export const getResourceAllocationAggregated: DetApi<
...
+  Api.V1QueryTrialsResponse
+> = {
+  name: 'queryTrials',
+  postProcess: (response: Api.V1QueryTrialsResponse): Api.V1QueryTrialsResponse => {
+    return response;
+  },
+  request: (params: Api.V1QueryTrialsRequest) => {
+    return detApi.TrialsComparison.queryTrials({
+      ...params,
+      limit: params?.limit ? 3 * params.limit : 30,
```
[the remark this answers is not in the document]
Comment (trentwatt) [reply webui/react/src/services/apiConfig.ts:454]: https://github.com/determined-ai/determined/blob/53ed8b324d37f76c8d184d3dfd14f041f15763b5/webui/react/src/pages/TrialsComparison/Table/TrialTable.tsx#L527-L541

this is in lieu of having a total for pagination (which would be costly for performance). it fetches the current page ) along with the next two pages- if they exist- and then we trim it down here `dataSource={trials.data.slice(0, settings.tableLimit)}`. 

so this is basically to know whether to know whether the next two pages exist so we know what `fakeTotal` to tell antd so that it will display the correct number of additional pages in the pagination control.

it's not pretty, and i think it would probably be better to implement backend side at some point but meh ¯\_(ツ)_/¯

File: webui/react/src/services/apiConfig.ts (hunk 3 above)
Comment (trentwatt) [reply webui/react/src/services/apiConfig.ts:507]: thanks!

File: webui/react/src/hooks/useHighlight.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  onRow: (record: RecordType) => {
+    onMouseEnter: (event: React.MouseEvent) => void;
+    onMouseLeave: (event: React.MouseEvent) => void;
+  };
+  rowClassName: (record: RecordType) => string;
+}
+
+type GetId<RecordType> = (record: RecordType) => number;
+
+function useHighlights<RecordType>(getId: GetId<RecordType>): Highlights<RecordType> {
```
Comment (ClaireNeveu) [inline webui/react/src/hooks/useHighlight.ts:19]: Can you add some documentation about what the purpose of this hook is?

File: webui/react/src/hooks/useModal/Experiment/useModalExperimentMove.tsx [hunk 6]
```diff
@@ -43,6 +43,7 @@ export interface Settings {
 }
 
 export const settingsConfig: SettingsConfig = {
+  applicableRoutespace: '/experiments',
```
Comment (ClaireNeveu) [inline webui/react/src/hooks/useModal/Experiment/useModalExperimentMove.tsx:46]: What's the purpose of this change?

**Review by ClaireNeveu — CHANGES_REQUESTED** [review]
Your test plan should be much more explicit and include all the things you are testing. This is a far-reaching PR so the test plan should be pretty big.

File: webui/react/src/hooks/useModal/Experiment/useModalExperimentMove.tsx (hunk 6 above)
Comment (trentwatt) [reply webui/react/src/hooks/useModal/Experiment/useModalExperimentMove.tsx:46]: specifying `applicableRoutespace` in general prevents [errors caused by settings interference](https://determinedai.atlassian.net/browse/WEB-490)-- `destroyInactivePane` by itself is not sufficient, at least with current implementation--  and imo is good practice to do whenever possible.

this modal is accessible from ProjectDetails -> Experiments tab and also the experiment details.

before this PR, the former existed at `/projects/{id}/`, but now it exists at `projects/{id}/experiments`, so now both pages have `/experiments` in their routes and it is practicable to specify.

Comment (trentwatt) [reply webui/react/src/hooks/useModal/Experiment/useModalExperimentMove.tsx:46]: if you say this `applicableRoutespace` is opaque and prone to bugs, i would agree with you

File: webui/react/src/components/Table/InteractiveTable.tsx [hunk 7]
```diff
@@ -305,19 +309,22 @@ const HeaderCell = ({
...
-          const minWidth = onResize(e, data);
+          onResize(e, data);
           const newWidth = data.x < minWidth ? minWidth : data.x;
 
           if (newWidth !== xValue) setXValue(newWidth);
         }}
         onStart={(e, data) => {
+          setShadowVisibility('block');
+          const newWidth = data.x < minWidth ? minWidth : data.x;
+          if (newWidth !== xValue) setXValue(newWidth);
```
Comment (trentwatt) [inline webui/react/src/components/Table/InteractiveTable.tsx:327]: the changes in this component are to fix the bug (currently in master) where the resizer is in the wrong place at the start

Uploading Screen Recording 2022-10-12 at 3.09.37 AM.mov…

File: webui/react/src/components/DynamicTabs.tsx [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+};
+
+export default DynamicTabs;
+
+export const useSetDynamicTabBar = (content: JSX.Element): void => {
+  const updateTabBarContent = useContext(TabBarContext);
+  useEffect(() => {
+    updateTabBarContent?.(content);
+  }, [updateTabBarContent, content]);
+};
```
Comment (trentwatt) [inline webui/react/src/components/DynamicTabs.tsx:77]: allows us to get rid of duplicated logic around tab routing.

also provides an alternative way to set the contents of the tab bar from within a tab itself. the previous implementation was to have all the code for the tab content running in the parent component of the tabgroup

File: webui/react/src/pages/ExperimentList.tsx [hunk 9]
```diff
@@ -0,0 +1,986 @@
+import { ExclamationCircleOutlined } from '@ant-design/icons';
```
Comment (trentwatt) [inline webui/react/src/pages/ExperimentList.tsx:1]: this is just the code for the `experiment` tab of `ProjectDetails`

File: webui/react/src/pages/ProjectNotes.tsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
+import { Button } from 'antd';
```
Comment (trentwatt) [inline webui/react/src/pages/ProjectNotes.tsx:1]: this is just the code for the `notes` of `ProjectDetails`

File: webui/react/src/components/Table/InteractiveTable.tsx (hunk 7 above)
Comment (ClaireNeveu) [reply webui/react/src/components/Table/InteractiveTable.tsx:328]: Can you put this in its own PR since it's unrelated to this?

File: webui/react/src/components/DynamicTabs.tsx (hunk 8 above)
Comment (ClaireNeveu) [reply webui/react/src/components/DynamicTabs.tsx:77]: Is this necessary for this feature or could it be moved to its own PR?

Comment (trentwatt) [reply webui/react/src/components/DynamicTabs.tsx:77]: this one in particular is needed to avoid having the experiment specific logic run in the parent component -- and therefore also when the experiment tab is inactive

Comment (trentwatt) [reply webui/react/src/components/DynamicTabs.tsx:77]: the difference between this and the other instances where we have different tables in different tabs is that this

image

has to be specified as a prop on `Tabs` itself

File: webui/react/src/components/Table/InteractiveTable.tsx (hunk 7 above)
Comment (trentwatt) [reply webui/react/src/components/Table/InteractiveTable.tsx:328]: will do.

File: webui/react/src/components/DynamicTabs.tsx [hunk 11]
```diff
@@ -0,0 +1,77 @@
+import { Tabs, TabsProps } from 'antd';
```
Comment (trentwatt) [inline webui/react/src/components/DynamicTabs.tsx:1]: not used unless feature flat is on now

File: webui/react/src/pages/TrialsComparison/Collections/useModalCreateCollection.tsx [hunk 12]
```diff
@@ -0,0 +1,147 @@
+import { Form, Input } from 'antd';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Collections/useModalCreateCollection.tsx:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Collections/useModalRenameCollection.module.scss [hunk 13]
```diff
@@ -0,0 +1,10 @@
+.base {
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Collections/useModalRenameCollection.module.scss:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Collections/useModalRenameCollection.tsx [hunk 14]
```diff
@@ -0,0 +1,99 @@
+import { Input } from 'antd';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Collections/useModalRenameCollection.tsx:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Collections/useModalViewFilters.tsx [hunk 15]
```diff
@@ -0,0 +1,110 @@
+import { Form } from 'antd';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Collections/useModalViewFilters.tsx:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Collections/useTrialCollections.tsx [hunk 16]
```diff
@@ -0,0 +1,397 @@
+import { Button, Dropdown, Menu, Select, Tooltip } from 'antd';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Collections/useTrialCollections.tsx:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Metrics/useLearningCurveData.ts [hunk 17]
```diff
@@ -0,0 +1,73 @@
+import { useCallback, useEffect, useState } from 'react';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Metrics/useLearningCurveData.ts:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Metrics/useMetricView.tsx [hunk 18]
```diff
@@ -0,0 +1,108 @@
+import React, { ReactNode, useCallback, useEffect, useReducer, useState } from 'react';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Metrics/useMetricView.tsx:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Table/TrialTable.tsx [hunk 19]
```diff
@@ -0,0 +1,557 @@
+import { Button, Dropdown } from 'antd';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Table/TrialTable.tsx:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Table/rangeFilter.tsx [hunk 20]
```diff
@@ -0,0 +1,64 @@
+import { FilterDropdownProps } from 'antd/lib/table/interface';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Table/rangeFilter.tsx:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Table/settings.ts [hunk 21]
```diff
@@ -0,0 +1,56 @@
+import { BaseType, SettingsConfig } from 'hooks/useSettings';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Table/settings.ts:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Table/tagsRenderer.tsx [hunk 22]
```diff
@@ -0,0 +1,66 @@
+import React, { ReactNode, useEffect, useState } from 'react';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Table/tagsRenderer.tsx:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Trials/data.ts [hunk 23]
```diff
@@ -0,0 +1,85 @@
+import { V1AugmentedTrial, V1QueryTrialsResponse } from 'services/api-ts-sdk';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Trials/data.ts:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/Trials/useFetchTrials.ts [hunk 24]
```diff
@@ -0,0 +1,51 @@
+import { useCallback, useEffect, useState } from 'react';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Trials/useFetchTrials.ts:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/TrialsComparison.tsx [hunk 25]
```diff
@@ -0,0 +1,107 @@
+import yaml from 'js-yaml';
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/TrialsComparison.tsx:1]: behind feature flag

File: webui/react/src/pages/TrialsComparison/api.ts [hunk 26]
```diff
@@ -0,0 +1,158 @@
+import {
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/api.ts:1]: behind feature flag

**Review by trentwatt — COMMENTED** [review]
put pretty much everything, including changes to existing code, behind a feature flag.

**Review by ClaireNeveu — APPROVED** [review]
Looks good. Make sure to go through your test plan before merging.

**Comment by trentwatt** [comment]
Tests (except for downstream test which is currently broken from stylelint PR) passed [here](https://github.com/determined-ai/determined/pull/4820/checks?sha=0a9b490b402a1f3b553e26b6fa41f0a13b51ffd5) with only trivial changes since.
